### ISSUE TITLE
`decouple_constraints` more efficient and more generic

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1007]
`decouple_constraints` currently is implemented in a very simple way. It first applies the whole standard update for all variables and then it applies another update for the constraints. However, this second update waits for all the first updates to finish, and this is not per parameter but for the whole model. Specifically:
```python
with tf_compat.v1.variable_scope("optimize_constraints"):
  with tf_compat.v1.variable_scope("factor"):
    factor = (self.get_current_step_learning_rate() / float(self.initial_learning_rate))
    factor *= self.config.float("decouple_constraints_factor", 0.025)
  sgd_optimizer = tf_compat.v1.train.GradientDescentOptimizer(
    learning_rate=factor, use_locking=self.use_locking)
  with tf.control_dependencies([self.optim_op]):
    self.optim_op = sgd_optimizer.minimize(self.constraints, var_list=self.trainable_vars)
```
`optim_op` before this was the single grouped op with all the updates.

This can be more efficient by using the `tf.Variable` `constraint` option which is specifically made for this purpose, and basically is the same but more efficient, per parameter:
```python
class _DenseResourceVariableProcessor(_OptimizableVariable):
  ...
  def update_op(self, optimizer, g):
    ...
    update_op = optimizer._resource_apply_dense(g, self._v)
    if self._v.constraint is not None:
      with ops.control_dependencies([update_op]):
        return self._v.assign(self._v.constraint(self._v))
    else:
      return update_op
```

Further, we could extend RETURNN to allow for other custom `constraint`.

**Comment by albertz** [comment]
Note that the TFA implementation of decoupled weight decay does actually the reverse order, i.e. it first applies weight decay, and then the gradient update. See [here](https://github.com/tensorflow/addons/blob/8cec33fcaaf1cf90aec7bdd55a0fcdbb251ce5c2/tensorflow_addons/optimizers/weight_decay_optimizers.py#L24). This should be more efficient.
